### PR TITLE
Add typescript definition

### LINKED ISSUE #47: Typescript definitions
I have created Typescript definitions, and wanted to check if you wanted to include this in the package or if I should simply add them to DefinitelyTyped.

```ts
declare module "react-native-app-auth" {
  interface IAppAuthProperties {
    scopes: string[];
    issuer: string;
    clientId: string;
    redirectUrl: string;
    aditionalParameters?: object;
  }

  function authorize(
    properties: IAppAuthProperties
  ): Promise<{
    accessToken: string;
    accessTokenExpirationDate: string;
    additionalParameters: object;
    idToken: string;
    refreshToken: string;
    tokenType: string;
  }>;

  function refresh(
    properties: IAppAuthProperties,
    { refreshToken: string }
  ): Promise<AuthorizeResult>;

  function revoke(
    properties: IAppAuthProperties,
    { tokenToRevoke: string, sendClientId: boolean }
  ): Promise<void>;
}
```

### PR DESCRIPTION
**Pull request by janpieterz**
Fixed #47 

First time adding definitions to a package, worked locally. Let me know if you want anything changed.

### CODE REVIEW

**Comment by jevakallio** [comment]
@kitten can you take a peek from TS good practices point of view when you have a moment. 

@kadikraman These seem legit to me, though I'd check if the optionality of fields is correct - IIRC the `revoke` call actually requires fewer parameters in the config?

Only thing I'd change is the name `AppAuthInput` to something like `AuthConfiguration`.

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  export function authorize(properties: AppAuthInput): Promise<AuthorizeResult>;
+
+  export function refresh(
+    properties: AppAuthInput,
+    { refreshToken: string }
+  ): Promise<AuthorizeResult>;
+
+  export function revoke(
+    properties: AppAuthInput,
+    { tokenToRevoke: string, sendClientId: boolean }
```
Comment (kadikraman) [inline index.d.ts]: - The only properties needed for `revoke` are `clientId` and `issuer`. It doesn't matter if you pass in the whole configuration object, but it's not all used. I'm not sure if there's an easy way to reflect that in TypeScript?
- `sendClientId` is optional, it defaults to `false`

Comment (janpieterz) [reply index.d.ts]: Thanks for the feedback!
As a proposal I made the `clientId` and `issues` in a separate interface that the `AuthConfiguration` extends, making it 1 definition. 
If we don't extend it'll mean we'd need to have two instances of the config object.

`sendClientId` made optional

**Comment by janpieterz** [comment]
Any idea for a name ( 😄 ) for the `clientId` and `issuer` revoke configuration different than `BaseAuthConfiguration`?

Changed name to `AuthConfiguration` as well.

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  export interface RefreshConfiguration {
+    refreshToken: string;
+  }
+
+  export function authorize(config: AuthConfiguration): Promise<AuthorizeResult>;
+
+  export function refresh(config: AuthConfiguration, refreshConfig: RefreshConfiguration): Promise<AuthorizeResult>;
+
+  export function revoke(config: BaseAuthConfiguration, refreshConfig: RevokeConfiguration): Promise<void>;
```
Comment (kadikraman) [inline index.d.ts]: Surely
```
revokeConfig: RevokeConfiguration
```

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
+export interface AuthConfiguration extends BaseAuthConfiguration {
+    scopes: string[];
+    redirectUrl: string;
+    aditionalParameters?: {[name: string]: any};
```
Comment (kadikraman) [inline index.d.ts]: Could you do
```
aditionalParameters?: {[name: string]: string};
```
as the values can technically only be strings?

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+  export interface RevokeConfiguration{
+    clientId: string;
+    issuer: string;
+  }
+
+  export interface AuthorizeResult {
+    accessToken: string;
+    accessTokenExpirationDate: string;
+    aditionalParameters?: {[name: string]: any};
```
Comment (kadikraman) [inline index.d.ts]: Same here

**Comment by janpieterz** [comment]
Done, cheers!

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
+export interface AuthConfiguration extends BaseAuthConfiguration {
```
Comment (kitten) [inline index.d.ts]: Minor nit here, the indentation seems to be off

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    scopes: string[];
+    redirectUrl: string;
+    aditionalParameters?: {[name: string]: string};
+  }
+
+  export interface BaseAuthConfiguration{
```
Comment (kitten) [inline index.d.ts]: can you move this to the top as `AuthConfiguration` extends it before it's being defined. Just for readability :+1:

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
+export interface AuthConfiguration extends BaseAuthConfiguration {
+    scopes: string[];
+    redirectUrl: string;
+    aditionalParameters?: {[name: string]: string};
```
Comment (kitten) [inline index.d.ts]: small typo: a**dd**itionalParameters

File: index.d.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  export interface BaseAuthConfiguration{
+    issuer: string;
+    clientId: string;
+  }
+
+  export interface RevokeConfiguration{
```
Comment (kitten) [inline index.d.ts]: can we make sure that there's a space before `{`?

@kadikraman: Can you point prettier—I think it's set up with lint-staged, right?—at all TS files as well?

File: index.d.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    clientId: string;
+    issuer: string;
+  }
+
+  export interface AuthorizeResult {
+    accessToken: string;
+    accessTokenExpirationDate: string;
+    aditionalParameters?: {[name: string]: string};
```
Comment (kitten) [inline index.d.ts]: same typo

**Comment by janpieterz** [comment]
Had disabled prettier locally because it kept massively reformatting the package.json on edit. 

Now uses prettier formatting and added *.ts to lint-staged too.

**Comment by PeterKottas** [comment]
Looks good to me! Gj

**Comment by jevakallio** [comment]
Looking good, landed this to master, will push it out in the next release. Thanks @janpieterz @kitten @kadikraman!

**Pull request merged**